Lib/multiprocessing/shared_memory.py

```diff
@@ -1,225 +1,316 @@
+        if not result:
+            raise ctypes.WinError(ctypes.get_last_error())
+        return args
+
+    kernel32.VirtualQuery.errcheck = _errcheck_bool
+    kernel32.VirtualQuery.restype = ctypes.c_size_t
+    kernel32.VirtualQuery.argtypes = (
+        wintypes.LPCVOID,
+        PMEMORY_BASIC_INFORMATION,
+        ctypes.c_size_t
+    )
+
+    kernel32.OpenFileMappingW.errcheck = _errcheck_bool
+    kernel32.OpenFileMappingW.restype = wintypes.HANDLE
+    kernel32.OpenFileMappingW.argtypes = (
+        wintypes.DWORD,
+        wintypes.BOOL,
+        wintypes.LPCWSTR
+    )
+
+    kernel32.CreateFileMappingW.errcheck = _errcheck_bool
+    kernel32.CreateFileMappingW.restype = wintypes.HANDLE
+    kernel32.CreateFileMappingW.argtypes = (
+        wintypes.HANDLE,
+        wintypes.LPCVOID,
+        wintypes.DWORD,
+        wintypes.DWORD,
+        wintypes.DWORD,
+        wintypes.LPCWSTR
+    )
+
+    kernel32.GetLastError.restype = wintypes.DWORD
+    kernel32.GetLastError.argtypes = ()
+
+    kernel32.MapViewOfFile.errcheck = _errcheck_bool
+    kernel32.MapViewOfFile.restype = wintypes.LPVOID
+    kernel32.MapViewOfFile.argtypes = (
+        wintypes.HANDLE,
+        wintypes.DWORD,
+        wintypes.DWORD,
+        wintypes.DWORD,
+        ctypes.c_size_t
+    )
+
+    kernel32.CloseHandle.errcheck = _errcheck_bool
+    kernel32.CloseHandle.argtypes = (wintypes.HANDLE,)
 
 
 class WindowsNamedSharedMemory:
 
-    def __init__(self, name, flags=None, mode=None, size=None, read_only=False):
+    def __init__(self, name, flags=None, mode=384, size=0, read_only=False):
         if name is None:
-            name = f'wnsm_{os.getpid()}_{random.randrange(100000)}'
+            name = _make_filename()
 
-        self._mmap = mmap.mmap(-1, size, tagname=name)
-        self.buf = memoryview(self._mmap)
+        if size == 0:
+            # Attempt to dynamically determine the existing named shared
+            # memory block's size which is likely a multiple of mmap.PAGESIZE.
+            try:
+                h_map = kernel32.OpenFileMappingW(PAGE_READONLY, False, name)
+            except OSError:
+                raise FileNotFoundError(name)
+            try:
+                p_buf = kernel32.MapViewOfFile(h_map, PAGE_READONLY, 0, 0, 0)
+            finally:
+                kernel32.CloseHandle(h_map)
+            mbi = MEMORY_BASIC_INFORMATION()
+            kernel32.VirtualQuery(p_buf, ctypes.byref(mbi), mmap.PAGESIZE)
+            size = mbi.RegionSize
+
+        if flags == O_CREX:
+            # Create and reserve shared memory block with this name until
+            # it can be attached to by mmap.
+            h_map = kernel32.CreateFileMappingW(
+                INVALID_HANDLE_VALUE,
+                None,
+                PAGE_EXECUTE_READWRITE,
+                size >> 32,
+                size & 0xFFFFFFFF,
+                name
+            )
+            try:
+                last_error_code = kernel32.GetLastError()
```

Comment:
This should call `ctypes.get_last_error()`, which is being called elsewhere. But for it to work, we need `use_last_error=True`, as mentioned above.

Reply:
Strike that. `kernel32.CreateFileMappingW._errcheck`  is set  to `_errcheck_bool`, which should raise `FileExistsError` here. Maybe this was added due to a misunderstanding. Using `ctypes.windll.kernel32` renders the `_errcheck` function useless, since `ctypes.get_last_error` isn't enabled.

Reply:
When ``kernel32.CreateFileMappingW`` is passed the name of an existing anonymous memory block in this way, it succeeds and it does not raise ``FileExistsError`` (``_errcheck_bool`` receives a non-zero ``result`` value).  This motivates the ``ctypes.get_last_error()`` call to check for a return value of ``183``.

This seems to match what is described on docs.microsoft.com, regarding the return value from ``CreateFileMappingA``:
```
    If the object exists before the function call, the function
    returns a handle to the existing object (with its current
    size, not the specified size), and GetLastError returns
    ERROR_ALREADY_EXISTS.
```

FWIW, this technique is leveraged in the Boost library for this same purpose.  The following is a snippet from boost/interprocess/windows_shared_memory.hpp where ``create_file_mapping`` is a wrapper for ``CreateFileMappingW``:
```
   switch(type){
      case ipcdetail::DoOpen:
         m_handle = winapi::open_file_mapping(map_access, filename);
      break;
      case ipcdetail::DoCreate:
      case ipcdetail::DoOpenOrCreate:
      {
         m_handle = winapi::create_file_mapping
            ( winapi::invalid_handle_value, protection, size, filename
            , (winapi::interprocess_security_attributes*)perm.get_permissions());
      }
      break;
      default:
         {
            error_info err = other_error;
            throw interprocess_exception(err);
         }
   }

   if(!m_handle || (type == ipcdetail::DoCreate && winapi::get_last_error() == winapi::error_already_exists)){
      error_info err = system_error_code();
      this->priv_close();
      throw interprocess_exception(err);
   }
```

I would be happier if ``CreateFileMappingW`` behaved as you suggested it might.

Reply:
> it succeeds and it does not raise `FileExistsError` (`_errcheck_bool` receives a non-zero `result` value). This motivates the `ctypes.get_last_error()` call to check for a return value of `183`.

Sorry it slipped my mind that Windows normally uses the [`OBJ_OPENIF`](https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/content/wudfwdm/ns-wudfwdm-_object_attributes) disposition when creating a named kernel object. If this flag weren't used, then the underlying `NtCreateSection` system call would fail with `STATUS_OBJECT_NAME_COLLISION`, which maps to Windows `ERROR_ALREADY_EXISTS`. Instead it succeeds, and `CreateFileMapping` manually sets the last error. It's one of a small number of cases in which a WINAPI call succeeds with an error set.